### DISCUSSION TITLE
Convention for floating base generalized coordinates

### DISCUSSION THREAD

**Discussion by FenglongSong** [stack-of-tasks/pinocchio#1657] [General]
Hi,

Thanks for the replies in https://github.com/stack-of-tasks/pinocchio/issues/1357 and https://github.com/stack-of-tasks/pinocchio/issues/1137, I got to know the generalized velocities are defined in local joint frames, which means for floating-base dynamics, the linear and angular velocities of the base are expressed in their own frames rather than world frame. Therefore, I'd like to confirm that is the generalized acceleration also expressed in the local frame of the joint?

That means, if I use `a = aba(model, data, q, v, tau)` to compute the generalized acceleration, `a` should be defined as:
`a = [linear_acceleration_in_body_frame, angular_acceleration_in_body_frame, joint_accelerations]`
Is that correct? I will appreciate your clarification! Thanks a lot.

**Comment by jcarpent** [comment]
Yes, you are correct. For the next question, I guess it would be better to use the discussion feature of GitHub to exchange on the topic. I think this might be of interest to others.

**Reply by FenglongSong** [reply]
Thanks! That's clear to me now :)

**Comment by craipy-hub** [comment]
Hello, @FenglongSong 
Thank you for the information you provided earlier.
I am using JointModelTranslation() and JointModelSphericalZYX() to create a floating base. From the previous discussion, I have learned the following information
```
q = [global_base_position, global_base_ZYX_euler_angles, joint_positions]
v = [local_base_velocity_linear, local_base_velocity_angular, joint_velocities]
```
But I am concerned about the local in v, There are still some questions about the expression of 'angular', I want to know local_base_velocity_angular is the order of angles
`[local_base_velocity_angular_x, local_base_velocity_angular_y, local_base_velocity_angular_z]`
, or 
`[local_base_velocity_angular_z, local_base_velocity_angular_y, local_base_velocity_angular_x]`?

**Reply by craipy-hub** [reply]
The right order of local_base_velocity_angular is 
``` 
local_base_velocity_angular = [local_base_velocity_angular_z, local_base_velocity_angular_y, local_base_velocity_angular_x]
```

**Reply by cmastalli** [reply]
Hi @binfengxueqian -- It is advisable to use `JointModelFreeFlyer` instead of your current solution. With this, you can avoid singularity associated with Euler angles.

**Reply by craipy-hub** [reply]
Okay, thank you. I will test `JointModelFreeFlyer` later

**Reply by craipy-hub** [reply]
> Hi @binfengxueqian -- It is advisable to use `JointModelFreeFlyer` instead of your current solution. With this, you can avoid singularity associated with Euler angles.

 @cmastalli Hi,
What is the general coordinates defined by `JointModelFreeFlyer`? Is it like the following?
```
q=[global_position_x,global_position_y,global_position_z,quaternion_x,quaternion_y,quaternion_z,quaternion_w]
v=[local_velocity_linear_x,local_velocity_linear_y,local_velocity_linear_z,local_velocity_angular_x,local_velocity_angular_y,local_velocity_angular_z]
```

**Reply by cmastalli** [reply]
The pose is defined through the SE3 convention. Instead, the twist is defined using the Featherstone convention, i.e., `local_vel_x, local_vel_y, local_vel_z`).

**Comment by manumerous** [comment]
Hi, after reading though the discussion above it is still unclear to me what the exact convention for the pinocchiovelocity vector is. 

1. Is it defined as follows? 
`v=[local_velocity_linear_x,local_velocity_linear_y,local_velocity_linear_z,local_velocity_angular_x,local_velocity_angular_y,local_velocity_angular_z, joint velocities]`

2. Is that true in both the case where we use the `JointModelFreeFlyer` or when we use `JointModelTranslation` + `JointModelSphericalZYX`?

**Reply by fabinsch** [reply]
Hi @manumerous ,

the generalized velocity vector is defined in the local coordinate frame of the joint.

If your joint is a JointModelFreeFlyer, you have `nq=7`, `nv=6`, where `v=[x_trans, y_trans, z_trans, x_rot, y_rot, z_rot]`.
If your joint is a JointModelSpherical**ZYX** (working with Euler angles), you have `nq=3`, `nv=3`, where `v=[z_rot, y_rot, x_rot]`.
If your joint is a JointModelSpherical (working with quaternions), you have `nq=4`, `nv=3`, where `v=[x_rot, y_rot, z_rot]`.

I hope this helps to clear up your questions.

**Reply by manumerous** [reply]
Thanks that helps a lot. Are [x_rot, y_rot, z_rot] always the components of the angular velocity? Or the derivatives of the euler angles in case of the `JointModelSphericalZYX`?

**Reply by fabinsch** [reply]
In the case of Euler angles, the angular velocity is the derivative of the Euler angles. 

You can verify this for example by checking how we integrate a configuration for this joint. This is handled by the `LieGroupMap` [here](https://github.com/stack-of-tasks/pinocchio/blob/master/include/pinocchio/multibody/liegroup/liegroup.hpp#L22), as the `JointModelSphericalZYX` does not appear in any of the special cases below (unlike `JointModelFreeFlyerTpl` or `JointModelSphericalTpl` or `JointModelRevoluteUnboundedTpl`). Then, you can check the `integrate_impl` of `VectorSpaceOperationTpl` [here](https://github.com/stack-of-tasks/pinocchio/blob/master/include/pinocchio/multibody/liegroup/vector-space.hpp#L126) which is just a simple addition.

**Reply by manumerous** [reply]
Thanks for the explanation. 

For the record, this phrase slightly confused me: "In the case of Euler angles, the angular velocity is the derivative of the Euler angles."
In general the angular velocity of a rigid body does not equal the derivative of the euler angles. I assume with the phrase above you do not mean angular velocity of a rigid body, but the angular velocity in the sense of the rotational velocity used as the v vector corresponding to that joint? 

I assume the jacobian would then also be computed with respect to the euler angle derivates instead of the angular velocity of that rigid body. At least it seems to be the case after looking through the code.
